mlx/ops.cpp

```diff
@@ -2897,4 +2897,27 @@ array tensordot(
   return reshape(matmul(x, y, s), rshape, s);
 }
 
+array outer(const array& a, const array& b, StreamOrDevice s /** = {}*/) {
+  if (a.ndim() != 1 || b.ndim() != 1) {
+    throw std::invalid_argument("[outer] a and b must be 1-dimensional.");
+  }
+
+  return multiply(reshape(a, {a.shape(0), 1}, s), b, s);
+}
+
+array inner(const array& a, const array& b, StreamOrDevice s /** = {}*/) {
+  if (a.dtype() != b.dtype()) {
+    throw std::invalid_argument("[inner] a and b must have the same dtype.");
+  }
```

Comment:
Must they have the same type? Why not promote them? (Also now that I see it, why don't we promote in `tensordot`?

Reply:
Yeah and since the ops being called handle this can just remove the checks all-together and pass-on. Will remove from tensordot as well.